improver/psychrometric_calculations/significant_phase_mask.py

```diff
@@ -0,0 +1,138 @@
+        """Initialise the class"""
+        self.lower_threshold = 0.01
+        self.upper_threshold = 0.99
+        self.phase_operator = {
+            "rain": self._rain_phase,
+            "sleet": self._sleet_phase,
+            "snow": self._snow_phase,
+        }
+
+    @staticmethod
+    def _validate_snow_fraction(snow_fraction):
+        """Ensures that the input snow-fraction field has appropriate
+          - name (snow_fraction)
+          - units (1)
+          - data (between 0 and 1 inclusive)
+
+        Raises
+            ValueError:
+                If any of the above are not True.
+        """
+        if snow_fraction.name() != "snow_fraction":
+            raise ValueError(
+                f"Expected cube named 'snow_fraction', not {snow_fraction.name()}"
+            )
+        if f"{snow_fraction.units}" != "1":
+            raise ValueError(f"Expected cube with units '1', not {snow_fraction.units}")
+        if (
+            isinstance(snow_fraction.data, np.ma.masked_array)
+            and snow_fraction.data.mask.any()
+        ):
+            raise NotImplementedError("SignificantPhaseMask cannot handle masked data")
+        if np.any((snow_fraction.data < 0) | (snow_fraction.data > 1)):
+            raise ValueError(
+                f"Expected cube data to be in range 0 <= x <= 1. "
+                f"Found max={snow_fraction.data.max()}; min={snow_fraction.data.min()}"
+            )
+
+    def _rain_phase(self, snow_fraction_data):
+        """Calculates the rain_phase data"""
+        return np.where(snow_fraction_data <= self.lower_threshold, 1, 0)
+
+    def _snow_phase(self, snow_fraction_data):
+        """Calculates the snow_phase data"""
+        return np.where(snow_fraction_data >= self.upper_threshold, 1, 0)
+
+    def _sleet_phase(self, snow_fraction_data):
+        """Calculates the sleet_phase data"""
+        return np.where(
+            (self.lower_threshold < snow_fraction_data)
+            & (snow_fraction_data < self.upper_threshold),
+            1,
+            0,
+        )
+
+    def process(self, snow_fraction, phase):
+        """
+        Make significant-phase-mask cube for the specified phase.
+
+        Args:
+            snow_fraction (iris.cube.Cube):
+                The input snow-fraction data to derive the phase mask from.
+            phase (str):
+                One of "rain", "sleet" or "snow". This is the phase mask that will be
+                returned.
+
+        Returns:
+            iris.cube.Cube:
+                The requested phase mask containing 1 where that phase is dominant
+                and 0 elsewhere. Dimensions will be identical to snow-fraction.
+        """
+        self._validate_snow_fraction(snow_fraction)
+
+        try:
+            data = self.phase_operator[phase](snow_fraction.data)
+        except KeyError:
+            raise KeyError(
+                f"Requested phase mask '{phase}' not in {self.phase_operator.keys()}"
```

Comment:
Codacy found an issue: [dict.keys referenced when not iterating](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=6368515)